source/partial-deep.d.ts

```diff
@@ -0,0 +1,27 @@
+/**
+Create a type from another type with all properties set to optional and the nested properties of all properties set to optional
+
+@example
+```
+import {PartialDeep} from 'type-fest';
+
+interface Baz {
+    qux: string;
+    corge: {
+        waldo: number[];
+        fred: string;
+    };
+}
+
+function foo(bar: PartialDeep<Baz>) { }
+
+foo({ corge: { waldo: [ 1, 2 ] } });
+```
+*/
+export type PartialDeep<T> = {
+    [KeyObject in keyof T]?: T[KeyObject] extends Array<infer InferredType>
+        ? Array<PartialDeep<InferredType>>
+        : T[KeyObject] extends ReadonlyArray<infer InferredType>
+            ? ReadonlyArray<PartialDeep<InferredType>>
+            : PartialDeep<T[KeyObject]>;
+};
```

Comment:
Does this also handle `Map` and `Set`? Would be good to have tests for those cases. See how it's handled in the other deep type: https://github.com/sindresorhus/type-fest/blob/5809b499720eb30f98967aedc8b9403a1427b26e/source/readonly-deep.d.ts#L32-L40

Reply:
Good point, handled in commit 92f9595

Reply:
Why didn't you just do the same as `ReadonlyDeep`? Why don't you handle primitive types and `Function`s, just like `ReadonlyDeep` (and all the other cases)?

We had a few iterations there, it's quite a solid definition.

Reply:
It does handle primitive types as showcased by tests. Regarding functions do you mean that it should convert

```TS
foo: { bar: { baz: function fn(_: string): boolean {} } }
```

to 

```TS
foo: { bar?: { baz?: function fn(_: string): boolean {} }
```

because it already does that, or do you mean that it should convert to

```TS
foo: { bar?: { baz?: function fn(_: string | undefined): boolean | null {} }
```

and if so then this type would indeed need more work. Probably would be wise to then also include Promisables if ReturnTypes are being taken into account.

Reply:
> It does handle primitive types as showcased by tests.

Yeah, I know it works with primitives the way it's defined right now, it's just a performance tweak because primitives are so common.

> to
> ```
> foo: { bar?: { baz?: function fn(_: string): boolean {} }
> ```
> because it already does that

Are you sure about this? There is no special handling for `Function`s. There is also no test for `Function`s. And it doesn't really work with `Function`s at all: https://www.typescriptlang.org/play/#code/KYDwDg9gTgLgBDAnmYcAKBDWBLDAbAEWGDAB4AVAPjgF44BvAKAEgBtAaWEXOVWwDs4Aay4QAZnHIBdAPwAuRnCXKl5Dlx4opcUDGD8AJgGc4AWQxkBY4FDMXO3XgBo4Vm3bAA1fAFdgm4EoWZhl0LBhcPHMyaIcAl2jvPD8AoOZmOUl1Ry0dED1DEwBlYBhSN1sSmFTg0MwcfCrSKpr0zLU43m1dfWM4AEEoKAxEcv5rW0HhnMDagaGR0nqI-CISUimR1LSMrM7cnsK4ACVgDAMIfjxETdGKk7OLq5uFmZ3Q0-PL69ul8Mi1mRPk8fq9tsFMssAcQyB0NF1KABuRgAX2RAj0UDEGAAxqgofhoqR9sAXEkUrxqIc+kSCYQYcT4ShKC46YDSOT-JTqPQUYwMTZsXiwg08E0qHkCn0mmyGVQeXzGIwAG5YOAYWgMMQQCBwTIACgAlLQFWilarbAAjPUilb09ZIFDidXUOgajAmWUO3jOjBIxiWgCEADptRAjcig6GdSGcfg8Pr+D48HhDYigA

Reply:
> because it already does that, or do you mean that it should convert to
> ```
> foo: { bar?: { baz?: function fn(_: string | undefined): boolean | null {}
> ```

No, I wouldn't do this.

Reply:
Hey, indeed it handled functions wrong and as I was adding more and more cases I realized that I should just move the definition to be in line with `ReadonlyDeep` as you suggested and that caused all the tests to pass, except now the negative test that I commented out doesn't raise an error.

Any ideas why that may be? The type is different from simply `Partial<T>`, but the type checker doesn't seem to think so.

Reply:
Okay, @BendingBender I seem to have fixed some issues that arose during rewriting the type to be more in line with `ReadonlyDeep`. Now all tests pass and the test suite covers all the same types that `ReadonlyDeep`s does.

I'd welcome some more feedback.